assert_writeln_magic.d

```diff
@@ -0,0 +1,313 @@
+    bool fromAssert;
+    bool fromStaticAssert;
+
+    Rebindable!(const AssertExpression) lastAssert;
+    Rebindable!(const EqualExpression) lastEqualExpression;
+
+    string fileName;
+    FileLines fl;
+}
+
+void parseFile(string fileName)
+{
+    import dparse.lexer;
+    import dparse.parser : parseModule;
+    import dparse.rollback_allocator : RollbackAllocator;
+    import std.array : uninitializedArray;
+
+    auto inFile = File(fileName);
+    if (inFile.size == 0)
+        warningf("%s is empty", inFile.name);
+
+    ubyte[] sourceCode = uninitializedArray!(ubyte[])(to!size_t(inFile.size));
+    inFile.rawRead(sourceCode);
+    LexerConfig config;
+    auto cache = StringCache(StringCache.defaultBucketCount);
+    const(Token)[] tokens = getTokensForParser(sourceCode, config, &cache).array;
+
+    RollbackAllocator rba;
+    auto m = parseModule(tokens, fileName, &rba);
+    auto visitor = new TestVisitor(fileName);
+    visitor.visit(m);
+    delete visitor;
+}
+
+void main(string[] args)
+{
+    import std.file;
+    import std.getopt;
+    import std.path : asNormalizedPath;
+
+    string inputDir;
+    string[] ignoredFiles;
+
+    auto helpInfo = getopt(args, config.required,
+            "inputdir|i", "Folder to start the recursive search for unittest blocks (can be a single file)", &inputDir,
+            "ignore", "List of files to exclude (partial matching is supported)", &ignoredFiles);
+
+    if (helpInfo.helpWanted)
+    {
+        return defaultGetoptPrinter(`assert_writeln_magic
+Tries to lower EqualExpression in AssertExpressions of Unittest blocks to commented writeln calls.
+`, helpInfo.options);
+    }
+
+    inputDir = inputDir.asNormalizedPath.array;
+
+    DirEntry[] files;
+
+    if (inputDir.isFile)
+    {
+        files = [DirEntry(inputDir)];
+        inputDir = ".";
+    }
+    else
+    {
+        files = dirEntries(inputDir, SpanMode.depth).filter!(
+                a => a.name.endsWith(".d") && !a.name.canFind(".git")).array;
+    }
+
+    foreach (file; files)
+        if (!ignoredFiles.any!(x => file.name.canFind(x)))
+            file.name.parseFile;
+}
+
+/**
+A simple line-based in-memory representation of a file.
+ - will automatically write all changes when the objct is destructed
+ - will use a temporary file to do safe, whole file swaps
+*/
+struct FileLines
+{
+    import std.array, std.file, std.path;
+
+    string[] lines;
+    string destFile, path, tmpDir;
+    bool hasWrittenChanges;
+
+    this(string path)
+    {
+        this(path.dirName, path.baseName);
+    }
+
+    this(string repoDir, string path)
+    {
+        writeln("opening: ", path);
+
+        this.path = path;
+        destFile = buildPath(repoDir, path);
+        lines = File(destFile).byLineCopy.array;
+
+        // it's a good practise to use a common tmp folder -> easier to look at or clean
+        tmpDir = buildPath(tempDir, "file_tester", path.stripExtension.replace("/", "_"));
```

Comment:
`tmpDir` is only used when writing the file line by line and doing the swap. It will be immediately be removed afterwards. Isn't this the dedicated purpose of a `tempDir()` or is there another reason to use `.generated`?